File: app/src/main/java/com/gmail/jorgegilcavazos/ballislife/features/games/GamesPresenter.java
package com.gmail.jorgegilcavazos.ballislife.features.games;

import com.gmail.jorgegilcavazos.ballislife.base.BasePresenter;
import com.gmail.jorgegilcavazos.ballislife.data.repository.games.GamesRepository;
import com.gmail.jorgegilcavazos.ballislife.features.model.GameV2;
import com.gmail.jorgegilcavazos.ballislife.util.DateFormatUtil;
import com.gmail.jorgegilcavazos.ballislife.util.GameUtils;
import com.gmail.jorgegilcavazos.ballislife.util.NetworkUtils;
import com.gmail.jorgegilcavazos.ballislife.util.schedulers.BaseSchedulerProvider;

import java.util.Calendar;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;

public class GamesPresenter extends BasePresenter<GamesView> {

    private final GamesRepository gamesRepository;
    private final BaseSchedulerProvider schedulerProvider;
    private final CompositeDisposable disposables = new CompositeDisposable();

    @Inject
    public GamesPresenter(
            GamesRepository gamesRepository, BaseSchedulerProvider schedulerProvider) {
        this.gamesRepository = gamesRepository;
        this.schedulerProvider = schedulerProvider;
    }

    public void onDateChanged() {
        view.hideGames();
    }

    public void loadModels(Calendar selectedDate, boolean forceNetwork) {
        view.dismissSnackbar();
        loadDateNavigatorText(selectedDate);

        disposables.clear();
        disposables.add(gamesRepository.models(selectedDate, forceNetwork)
                                       .observeOn(schedulerProvider.ui(), true)
                                       .subscribeWith(new DisposableObserver<GamesUiModel>() {
                                           @Override
                                           public void onNext(GamesUiModel uiModel) {
                                               if (uiModel.isMemorySuccess() && uiModel.getGames()
                                                                                       .isEmpty()) {
                                                   view.setLoadingIndicator(true);
                                               }

                                               if (uiModel.isNetworkSuccess()) {
                                                   view.setLoadingIndicator(false);
                                               }

                                               if (uiModel.getGames() != null && !uiModel.getGames()
                                                                                         .isEmpty()) {
                                                   view.showGames(uiModel.getGames());
                                               }

                                               view.setNoGamesIndicator(uiModel.isNetworkSuccess
                                                       () && uiModel
                                                       .getGames()
                                                       .isEmpty());
                                           }

                                           @Override
                                           public void onError(Throwable e) {
                                               view.setLoadingIndicator(false);
                                               if (NetworkUtils.Companion.isNetworkAvailable()) {
                                                   view.showErrorSnackbar();
                                               } else {
                                                   view.showNoNetSnackbar();
                                               }
                                           }

                                           @Override
                                           public void onComplete() {

                                           }
                                       }));
    }

    public void loadDateNavigatorText(Calendar selectedDate) {
        String dateText = DateFormatUtil.formatNavigatorDate(selectedDate.getTime());
        view.setDateNavigatorText(dateText);
    }

    public void openGameDetails(GameV2 requestedGame, Calendar selectedDate) {
        view.showGameDetails(requestedGame, selectedDate);
    }

    public void updateGames(String gameData, Calendar selectedDate) {
        if (DateFormatUtil.isDateToday(selectedDate.getTime())) {
            view.updateScores(GameUtils.getGamesListFromJson(gameData));
        }
    }

    public void stop() {
        disposables.clear();
        view.dismissSnackbar();
    }

}

File: app/src/main/java/com/gmail/jorgegilcavazos/ballislife/features/games/GamesPresenter.kt
package com.gmail.jorgegilcavazos.ballislife.features.games

import com.gmail.jorgegilcavazos.ballislife.base.BasePresenter
import com.gmail.jorgegilcavazos.ballislife.data.local.LocalRepository
import com.gmail.jorgegilcavazos.ballislife.features.model.GameV2
import com.gmail.jorgegilcavazos.ballislife.util.ErrorHandler
import com.gmail.jorgegilcavazos.ballislife.util.NetworkUtils
import com.gmail.jorgegilcavazos.ballislife.util.TeamUtils
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.OnErrorNotImplementedException
import io.reactivex.rxkotlin.addTo
import javax.inject.Inject

class GamesPresenter @Inject constructor(
    private val localRepository: LocalRepository,
    private val gamesModelTransformer: GamesModelTransformer,
    private val disposables: CompositeDisposable,
    private val networkUtils: NetworkUtils,
    private val errorHandler: ErrorHandler) : BasePresenter<GamesView>() {

  override fun attachView(view: GamesView) {
    super.attachView(view)

    val uiEvents = Observable.merge(
        view.loadGamesEvents(),
        view.refreshGamesEvents(),
        view.openGameEvents()
    )

    gamesModelTransformer.uiModels(uiEvents).subscribe({ model ->
      when (model) {
        is GamesUiModelV2.Idle -> {
          view.setNoGamesIndicator(false)
        }
        is GamesUiModelV2.MemoryInProgress -> {
          view.hideGames()
          view.setLoadingIndicator(false)
          view.setNoGamesIndicator(false)
          view.dismissSnackbar()
        }
        is GamesUiModelV2.NetworkInProgress -> {
          view.setNoGamesIndicator(false)
          view.dismissSnackbar()
        }
        is GamesUiModelV2.Success -> {
          if (view.getCurrentDateShown() == model.date) {
            view.showGames(gamesListWithFavoriteTeamFirst(model.games.toMutableList()))
            view.setLoadingIndicator(false)
            view.setNoGamesIndicator(false)
          }
        }
        is GamesUiModelV2.NoCachedGames -> {
          view.setLoadingIndicator(true)
          view.setNoGamesIndicator(false)
        }
        is GamesUiModelV2.NoGames -> {
          view.setNoGamesIndicator(true)
          view.setLoadingIndicator(false)
        }
        is GamesUiModelV2.Failure -> {
          view.setLoadingIndicator(false)
          if (networkUtils.isNetworkAvailable()) {
            view.showErrorSnackbar(errorHandler.handleError(model.t))
          } else {
            view.showNoNetSnackbar()
          }
          view.setNoGamesIndicator(false)
        }
        is GamesUiModelV2.OpenGameScreen -> {
          view.showGameDetails(model.game)
        }
      }
    }, { t ->
      throw OnErrorNotImplementedException(t)
    }).addTo(disposables)
  }

  override fun detachView() {
    disposables.clear()
    view.dismissSnackbar()
    super.detachView()
  }

  /**
   * Moves the user's favorite team to the top if one is selected.
   */
  private fun gamesListWithFavoriteTeamFirst(games: MutableList<GameV2>): List<GameV2> {
    val favTeam = localRepository.favoriteTeam
    if (favTeam == null) {
      // No favorite team selected, return unchanged list.
      return games
    } else {
      val favTeamId = TeamUtils.getTeamId(favTeam)
      val favTeamGame = games.find { it.homeTeamId == favTeamId || it.awayTeamId == favTeamId }
      return if (favTeamGame == null) {
        // Favorite team not in list of games, return unchanged list.
        games
      } else {
        // Remove favorite team from current position and place at index 0, then return that list.
        games.remove(favTeamGame)
        games.add(0, favTeamGame)
        games
      }
    }
  }

}
